File: 01guessthedate/solution.cpp
#include <iostream>
#include <ctime>


int main() {
    int year, month, day;
    const int longMonths[] = {1, 3, 5, 7, 8, 10, 12};

    std::cin >> year >> month >> day;

    day -= 1;
    auto checkDay = false;

    if(day < 1) {
        day = 31;
        month -= 1;
        checkDay = true;
    }
    if(month < 1) {
        month = 12;
        year -= 1;
    }

    if(checkDay) {
        if(month == 2) {
            if(year % 400 == 0) {
                day = 29;
            }
            else {
                day = 28;
            }
        }
        else {
            bool has30days = true;
            for(auto monthLength = longMonths; monthLength != longMonths + 7; monthLength++) {
                if(month == *monthLength) {
                    has30days = false;
                }
            }
            if(has30days) {
                day = 30;
            }
        }
    }

    std::cout << day << month << year << '\n';
    return 0;
}

File: 01guessthedate/solution.js
const getGets = (arr) => {
    let index = 0;

    return () => {
        const toReturn = arr[index];
        index += 1;
        return toReturn;
    };
};
// this is the test
const test = `2017
12
31`.split('\n');

const gets = this.gets || getGets(test);
const print = this.print || console.log;
const quit = (() => {});

const getMonthName = (index) => {
    const monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
    return monthNames[index]
};

const formatDate = (date) => {
    const yearString = date.getFullYear();
    const monthString = getMonthName(date.getMonth());
    // const monthString = date.getMonth();
    const dayString = date.getDate();
    return `${dayString}-${monthString}-${yearString}`;
};

const solve = () => {
    const year = gets();
    const month = gets();
    const day = gets();

    const date = new Date(year, month - 1, day);
    // print(formatDate(date));
    date.setDate(date.getDate() - 1);
    // print(formatDate(date));
    return formatDate(date);
}

print(solve());
